### PR TITLE
Alerting: Allow querying of Alerts from notifications

### PR DESCRIPTION
**Pull request by gotjosh**
Interface for querying the alerts currently held in the notification system. 

With this, we have the whole story for creating silences and then querying the alerts back.

- [x] Make changes to alerting-api for supporting this
- [x] Revise usage of locks
- [x] Revise and alias errors
- [x] Provide example requests for manual testing
- [x] Hook up API endpoints

**Special notes for your reviewer**:

### CODE REVIEW

File: pkg/tests/testinfra/testinfra.go [hunk 1]
```diff
@@ -195,6 +196,12 @@ func CreateGrafDir(t *testing.T, opts ...GrafanaOpts) (string, string) {
 			_, err = securitySect.NewKey("content_security_policy", "true")
 			require.NoError(t, err)
 		}
+		if len(o.EnableFeatureToggles) > 0 {
```
Comment (gotjosh) [inline pkg/tests/testinfra/testinfra.go:199]: This enables us to do feature toggling within the integration test. cc: @aknuds1 

The looping through options threw me off at the beginning but I guess it's a common idiom for optional params.

File: pkg/tests/api/alerting/api_alertmanager_test.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+)
+
+func TestAlertAndGroupsQuery(t *testing.T) {
+	dir, path := testinfra.CreateGrafDir(t, testinfra.GrafanaOpts{
+		EnableFeatureToggles: []string{"ngalert"},
+	})
+	store := setupDB(t, dir)
+	grafanaListedAddr := testinfra.StartGrafana(t, dir, path, store)
+
+	// When there are no alerts available, it returns an empty list.
```
Comment (gotjosh) [inline pkg/tests/api/alerting/api_alertmanager_test.go:25]: We currently have no way to introduce alerts via the API (it is deliberately disabled) and I didn't want to have to go through creating an alert and waiting for it to come up. 

For now, this is enough as we'll revisit this soon with the silences tests.

File: pkg/services/ngalert/api/test-data/am-alertmanager-recipient.http [hunk 3]
```diff
@@ -97,6 +97,10 @@ content-type: application/json
 # get AM alerts
 GET http://admin:admin@localhost:3000/alertmanager/{{alertManagerDatasourceID}}/api/v2/alerts
 
+###
+# get AM alert groups
+GET http://admin:admin@localhost:3000/alertmanager/{{alertManagerDatasourceID}}/api/v2/alerts/groups
```
Comment (papagian) [inline pkg/services/ngalert/api/test-data/am-alertmanager-recipient.http:102]:
```suggestion
GET http://admin:admin@localhost:3000/api/alertmanager/{{alertManagerDatasourceID}}/api/v2/alerts/groups
```

File: pkg/services/ngalert/api/test-data/am-alertmanager-recipient.http [hunk 4]
```diff
@@ -164,4 +168,4 @@ DELETE http://admin:admin@localhost:3000/alertmanager/{{alertManagerDatasourceID
 
 ###
 # delete silence - unknown
-DELETE http://admin:admin@localhost:3000/alertmanager/{{alertManagerDatasourceID}}/api/v2/silence/unknown
\ No newline at end of file
+DELETE http://admin:admin@localhost:3000/alertmanager/{{alertManagerDatasourceID}}/api/v2/silence/unknown
```
Comment (papagian) [inline pkg/services/ngalert/api/test-data/am-alertmanager-recipient.http:171]:
```suggestion
DELETE http://admin:admin@localhost:3000/api/alertmanager/{{alertManagerDatasourceID}}/api/v2/silence/unknown
```

File: pkg/tests/api/alerting/api_alertmanager_test.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
+package alerting
```
Comment (papagian) [inline pkg/tests/api/alerting/api_alertmanager_test.go:1]: Alternatively, we could have this test under [pkg/services/ngalert/tests](https://github.com/grafana/grafana/tree/master/pkg/services/ngalert/tests)

Comment (gotjosh) [reply pkg/tests/api/alerting/api_alertmanager_test.go:1]: I originally tried that, but given testinfra is a top-level package it wouldn't let me due to circular dependencies.

File: pkg/tests/api/alerting/api_alertmanager_test.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+		resp, err := http.Get(alertsURL)
+		require.NoError(t, err)
+		t.Cleanup(func() {
+			err := resp.Body.Close()
+			require.NoError(t, err)
+		})
+		b, err := ioutil.ReadAll(resp.Body)
+		require.NoError(t, err)
+		require.NoError(t, err)
+		require.Equal(t, 200, resp.StatusCode)
```
[the remark this answers is not in the document]
Comment (gotjosh) [reply pkg/tests/api/alerting/api_alertmanager_test.go:37]: I don't mind do it if I get to it first. Is it OK if we do it in the next PR?

I also noticed this but thought it wasn't a problem until rule creation as it's the only part that has explicit requirements on this.

[the remark this answers is not in the document]
Comment (papagian) [reply pkg/tests/api/alerting/api_alertmanager_test.go:37]: yes, go ahead!

**Review by papagian — APPROVED** [review]

**Pull request merged**